dpnp/backend/extensions/lapack/gesv_batch.cpp

```diff
@@ -0,0 +1,326 @@
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//*****************************************************************************
+
+#include <pybind11/pybind11.h>
+
+// dpctl tensor headers
+#include "utils/memory_overlap.hpp"
+#include "utils/output_validation.hpp"
+#include "utils/type_utils.hpp"
+
+#include "common_helpers.hpp"
+#include "gesv.hpp"
+#include "linalg_exceptions.hpp"
+#include "types_matrix.hpp"
+
+#include "dpnp_utils.hpp"
+
+namespace dpnp::extensions::lapack
+{
+namespace mkl_lapack = oneapi::mkl::lapack;
+namespace py = pybind11;
+namespace type_utils = dpctl::tensor::type_utils;
+
+typedef sycl::event (*gesv_batch_impl_fn_ptr_t)(
+    sycl::queue &,
+    const std::int64_t,
+    const std::int64_t,
+    const std::int64_t,
+    char *,
+    char *,
+    const std::vector<sycl::event> &);
+
+static gesv_batch_impl_fn_ptr_t
+    gesv_batch_dispatch_vector[dpctl_td_ns::num_types];
+
+template <typename T>
+static sycl::event gesv_batch_impl(sycl::queue &exec_q,
+                                   const std::int64_t n,
+                                   const std::int64_t nrhs,
+                                   const std::int64_t batch_size,
+                                   char *in_a,
+                                   char *in_b,
+                                   const std::vector<sycl::event> &depends)
+{
+    type_utils::validate_type_for_device<T>(exec_q);
+
+    T *a = reinterpret_cast<T *>(in_a);
+    T *b = reinterpret_cast<T *>(in_b);
+
+    std::int64_t a_size = n * n;
+    std::int64_t b_size = n * nrhs;
+
+    const std::int64_t lda = std::max<size_t>(1UL, n);
+    const std::int64_t ldb = std::max<size_t>(1UL, n);
+
+    // Get the number of independent linear streams
+    std::int64_t n_linear_streams =
+        (batch_size > 16) ? 4 : ((batch_size > 4 ? 2 : 1));
+
+    const std::int64_t scratchpad_size =
+        mkl_lapack::gesv_scratchpad_size<T>(exec_q, n, nrhs, lda, ldb);
+
+    T *scratchpad =
+        helper::alloc_scratchpad<T>(scratchpad_size, n_linear_streams, exec_q);
+
+    // Get padding size to ensure memory allocations are aligned to 256 bytes
+    // for better performance
+    const std::int64_t padding = 256 / sizeof(T);
+
+    // Calculate the total size needed for the pivot indices array for all
+    // linear streams with proper alignment
+    size_t alloc_ipiv_size =
+        helper::round_up_mult(n_linear_streams * n, padding);
+
+    // Allocate memory for the total pivot indices array
+    std::int64_t *ipiv =
+        sycl::malloc_device<std::int64_t>(alloc_ipiv_size, exec_q);
+    if (!ipiv)
+        throw std::runtime_error("Device allocation for ipiv failed");
```

Comment:
Should we created `alloc_ipiv` like we did for scratchpad?

Reply:
Sounds good. I have added `alloc_ipiv` to `common_helpers.hpp`. Thanks